lib/attr_encrypted.rb

```diff
@@ -324,35 +370,53 @@ def evaluate_attr_encrypted_option(option)
         end
       end
 
-      def load_iv_for_attribute(attribute, algorithm)
-        encrypted_attribute_name = self.class.encrypted_attributes[attribute.to_sym][:attribute]
+      def load_iv_for_attribute(attribute, options)
+        encrypted_attribute_name = options[:attribute]
+        encode_iv = options[:encode_iv]
         iv = send("#{encrypted_attribute_name}_iv")
-        if(iv == nil)
+        if options[:operation] == :encrypting
           begin
-            algorithm = algorithm || "aes-256-cbc"
-            algo = OpenSSL::Cipher::Cipher.new(algorithm)
-            iv = [algo.random_iv].pack("m")
+            iv = generate_iv(options[:algorithm])
+            iv = [iv].pack(encode_iv) if encode_iv
             send("#{encrypted_attribute_name}_iv=", iv)
           rescue RuntimeError
           end
         end
-        self.class.encrypted_attributes[attribute.to_sym] = self.class.encrypted_attributes[attribute.to_sym].merge(:iv => iv.unpack("m").first) if (iv && !iv.empty?)
+        if iv && !iv.empty?
+          iv = iv.unpack(encode_iv).first if encode_iv
+          options[:iv] = iv
+        end
+      end
+
+      def generate_iv(algorithm)
+        algo = OpenSSL::Cipher.new(algorithm)
+        algo.encrypt
+        algo.random_iv
       end
 
-      def load_salt_for_attribute(attribute)
-        encrypted_attribute_name = self.class.encrypted_attributes[attribute.to_sym][:attribute]
-        salt = send("#{encrypted_attribute_name}_salt") || send("#{encrypted_attribute_name}_salt=", generate_random_base64_encoded_salt)
-        salt = decode_salt_if_encoded(salt)
-        self.class.encrypted_attributes[attribute.to_sym] = self.class.encrypted_attributes[attribute.to_sym].merge(:salt => salt)
+      def load_salt_for_attribute(attribute, options)
+        encrypted_attribute_name = options[:attribute]
+        encode_salt = options[:encode_salt]
+        salt = send("#{encrypted_attribute_name}_salt")
+        if (salt == nil)
+          salt = SecureRandom.random_bytes
+          salt = prefix_and_encode_salt(salt, encode_salt) if encode_salt
+          send("#{encrypted_attribute_name}_salt=", salt)
+        end
+        if salt && !salt.empty?
+          salt = decode_salt_if_encoded(salt, encode_salt) if encode_salt
+          options[:salt] = salt
+        end
       end
 
-      def generate_random_base64_encoded_salt
+      def prefix_and_encode_salt(salt, encoding)
         prefix = '_'
-        prefix + [SecureRandom.random_bytes].pack("m")
+        prefix + [salt].pack(encoding)
       end
 
-      def decode_salt_if_encoded(salt)
-        salt.slice(0).eql?('_') ? salt.slice(1..-1).unpack("m").first : salt
+      def decode_salt_if_encoded(salt, encoding)
+        prefix = '_'
```

Comment:
I see this used in `prefix_and_encode_salt`, do you think it should be a constant?